### DISCUSSION TITLE
EditObjectDialog Relationship to other Entity

### DISCUSSION THREAD

**Discussion by Quotic** [fgilde/MudBlazor.Extensions#60] [Q&A] [answered]
Hi @fgilde,

again a question from my side.. I tried but unfortunately without success...

I have a class which I want to create an object with EditObject Dialog..

```
    public class Profile : BaseEntity
    {
        [Required]
        public string Name { get; set; }
        public string Description { get; set; } = string.Empty;
        public bool IsActive { get; set; } = true;
        public virtual ICollection<Folder>? Folders { get; set; }
    }
```

But for Folders Collection I want a select or list where I can select already created Folders to add to this Profile object.. Is it possible to overwrite the objecteditmeta of this FolderProperty with a Dictionary or List, where User can select one or more of the already created Folders?

I hope you understand what I mean..

Thanks

Bastian

**Comment by fgilde** [comment, marked as answer]
Hi, I've noticed you're from Germany, so you can write in German if you prefer—I'm from Hamburg. However, since you've asked your question in English, I'll respond in English.

Yes, you can create complex UIs using the ObjectEditMeta. In one of my projects, I used it for complex tree node selection involving multiple items and hierarchical data.

I've created a sample for you here: https://trymudex.azurewebsites.net/snippet/GOmdPvEAcCodvJSZ

I'm not familiar with your structure in detail, but for a simple selection from a `List<Folder>`, you can use `RenderWith` and specify any control you want. You could also directly use `MudExSelect<Folder>` and define all properties as shown in the code below, but I would recommend using a custom control—for instance, one that inherits from [MudExSelect](https://www.mudex.org/mud-ex-select).

```c#
    private Action<ObjectEditMeta<Profile>> Configure()
    {
        return meta =>
        {
            meta.Property(m => m.Folders)
                .RenderWith<MudExSelect<Folder>, ICollection<Folder>, IEnumerable<Folder>>(fs => fs.SelectedValues, select =>
                {
                    select.ToStringFunc = folder => folder?.Name;
                    select.MultiSelection = true;
                    select.AvailableItemsLoadFunc = token => FolderService.GetAllFoldersAsync(token); // Your service to retrieve Items
                    ....
                    ....
                })         
        };
    }
```

You can then configure it in the `MudExObjectEdit` like this:
`<MudExObjectEdit MetaConfiguration="@Configure()" Value="..."></MudExObjectEdit>`
Or specify it in a dialog edit manner like this:
`await dialogService.EditObject(new Profile(), "Title", OnSubmitCallback, DialogOptionsEx, Configure());`

In the sample I made for you, I've created a `FolderSelect` component that's used.

Sample link: https://trymudex.azurewebsites.net/snippet/GOmdPvEAcCodvJSZ

One more note: If you want to render a profile in the same manner across multiple pages or dialogs, you can implement `IObjectMetaConfiguration<Profile>`, which will then be automatically registered, so you won't need to specify the `Configure` method as mentioned above.


```c#
    public class ProfileMetaConfig : IObjectMetaConfiguration<Profile>
    {
        public Task ConfigureAsync(ObjectEditMeta<Profile> meta)
        {
            meta.Property(m => m.Folders)
                .RenderWith<MudExSelect<Folder>, ICollection<Folder>, IEnumerable<Folder>>(fs => fs.SelectedValues, select =>
                {
                    select.AvailableItemsLoadFunc = token => FolderService.GetAllFoldersAsync(token);
                });
            return Task.CompletedTask;
        }
    }
```

And if you want to always render folders with a specific control, regardless of the model, you can use the `RenderDataDefaults`. For example, in your case:

```c#
RenderDataDefaults.RegisterDefault<ICollection<Folder>, IEnumerable<Folder>, FolderSelect>(s => s.SelectedValues); // all ICollection<Folder> are rendered with FolderSelect
RenderDataDefaults.RegisterDefault<Folder[], IEnumerable<Folder>, FolderSelect>(s => s.SelectedValues); // all arrays of Folder are rendered with FolderSelect

```

Hope that helps

**Comment by Quotic** [comment]
Yes, I have already noticed that. Greetings from Bavaria :) My hope is that my question and your very detailed answer might also help others. For this reason, I always ask in English. I am a big fan of your extensions, which make my work (private projects, unfortunately not my main job :)) much easier.

**Reply by fgilde** [reply]
Thank you very much

**Reply by Quotic** [reply]
Jetzt in Deutsch, da ich schon etwas fertig bin ;) 
![image](https://github.com/fgilde/MudBlazor.Extensions/assets/1618550/cbcb03e7-1d14-4c7f-9d73-775922dbddd3)

Aus irgendeinem Grund zeigt er mir die ausgewählten Werte nicht an. Die Auflösung der "Folder" passt:
![image](https://github.com/fgilde/MudBlazor.Extensions/assets/1618550/f15f321c-5c57-4d88-bb1c-09b10d80250f)

Wenn ich es auf 
'''
        ValuePresenter = ValuePresenter.Chip;
'''
umstelle zeigt er mir aber die Werte an und im gespeicherten Objekt sind sie auch mit drin. Habe der Einfachheit alles wie in deinem Beispiel gemacht. Bei meinem Versuch das ganze auch ohne MultiSelect zu machen, zeigt er mir auch die Auswahl an, aber wenn ich etwas ausgewählt habe, nimmt er es in der Form nicht an und es wird auch nicht dem Profil Objekt hinzugefügt... Hast du eine Idee woran das liegen könnte?

**Comment by fgilde** [comment]
Kannst du ne sample solution erstellen? Dann kann ich mir das gerne mal ansehen. Kann generell mehrere Gründe haben. 
`ValuePresenter.Text`  sollte in jedem fall funktionieren `ValuePresenter.None` würde dafür sorgen dass nichts angezeigt wird. Wenn da das problem liegt. 
Aber sonst kannst du in MudExSelect bzw deiner ableitung auch ``ValuePresenter.ItemContent` nehmen und selber ein Item Template erstellen. 

```c#
    protected override async Task OnInitializedAsync()
    {        
        ValuePresenter = ValuePresenter.ItemContent;
        ItemTemplate = GetItemTemplate();
        await base.OnInitializedAsync();
    }
    private RenderFragment<Branch> GetItemTemplate() => context =>
        @<div style="display: flex; height: 25px;">
            <div style=@($"margin:5px; width:15px; height:15px; border-radius:50%; background-color: #{context.Color}")></div>
            <p style="margin: 5px 0 0 3px">@context.Title</p>
        </div>;
```

**Reply by Quotic** [reply]
Tatsächlich hat es mit dem ItemTemplate funktioniert, aber es ist wohl mehr im Argen, da beim erneuten öffnen der Dropdown Box die ausgewählten Einträge nicht mehr markiert sind. Ich versuche heute Abend mal eine Sample solution zu erstellen.

**Reply by Quotic** [reply]
@fgilde also ich versuche es gerade mit einer SampleSolution zu reproduzieren... aber tatsächlich funktioniert es dort, also ist es wahrscheinlich irgendein Seiteneffekt... Ich werde dir feedback geben, wenn ich es herausgefunden habe.

**Reply by Quotic** [reply]
Aber was ich tatsächlich nicht hinbekomme. Wie ist es denn wenn ich MultiSelect deaktiviere? Normalerweise wird ja trotzdem ein IEnumerable zurückgegeben welches nur ein element enthält oder?

**Reply by fgilde** [reply]
Das ist zwar korrekt, du solltest dann aber `Value` nehmen statt SelectedValues.
Falls du noch probleme mit dem ValuePresenter.Text hast nimm mal das previewe package [1.7.77-prev-2311121951-main](https://www.nuget.org/packages/MudBlazor.Extensions/1.7.77-prev-2311121951-main) hab da ne kleinigkeit geändert.
